rfc/src/rfcs/0005-should-fail-attr.md

```diff
@@ -0,0 +1,250 @@
+
+## User Impact
+
+Users may want to express that a verification harness should panic.
+In general, a user adding such a harness wants to demonstrate that the verification fails because a panic is reachable from the harness.
+
+Let's refer to this concept as *negative verification*,
+so the relation with [negative testing](https://en.wikipedia.org/wiki/Negative_testing) becomes clearer.
+Negative testing can be exercised in Rust unit tests using [the `#[should_panic]` attribute](https://doc.rust-lang.org/rust-by-example/testing/unit_testing.html#testing-panics).
+If the `#[should_panic]` attribute is added to a test, `cargo test` will check that the execution of the test results in a panic.
+This capability doesn't exist in Kani at the moment, but it would be useful for the same reasons
+(e.g., to show that invalid inputs result in verification failures, or increase the overall verification coverage).
+
+We propose an attribute that allows users to exercise negative verification in Kani.
+
+We also acknowledge that, in other cases, users may want to express more granular expectations for their harnesses.
+For example, a user may want to specify that a given check is unreachable from the harness.
+An ergonomic mechanism for informing Kani about such expectations is likely to require other improvements in Kani (a comprehensive classification for checks reported by Kani, a language to describe expectations for checks and cover statements, and general output improvements).
+Moving forward, we consider that such a mechanism and this proposal solve different problems, so they don't need to be discussed together.
+This is further discussed in the [rationale and alternatives](#rationale-and-alternatives) and [future possibilities](#future-possibilities) sections.
+
+## User Experience
+
+The scope of this functionality is **limited to the overall verification result**.
+The [rationale section](#rationale-and-alternatives) discusses the granularity of failures, and how this attribute could be extended.
+
+### Single Harness
+
+Let's look at this code:
+
+```rust
+struct Device {
+    is_init: bool,
+}
+
+impl Device {
+    fn new() -> Self {
+        Device { is_init: false }
+    }
+
+    fn init(&mut self) {
+        assert!(!self.is_init);
+        self.is_init = true;
+    }
+}
+
+#[kani::proof]
+fn cannot_init_device_twice() {
+    let mut device = Device::new();
+    device.init();
+    device.init();
+}
+```
+
+This is what a negative harness may look like.
+The user wants to verify that calling `device.init()` more than once should result in a panic.
+
+> **NOTE**: We could convert this into a Rust unit test and add the `#[should_panic]` attribute to it.
+> However, there are two good reasons to have a verification-specific attribute that does the same:
+>  1. To ensure that other unexpected behaviors don't occur (e.g., overflows).
+>  2. Because `#[should_panic]` cannot be used if the test harness contains calls to Kani's API.
```

Comment:
A third potential case is that the user has used `cfg` attributes to stub out part of the code, and wants to be sure the expected panic occurs both in the code as exposed to Kani and in the "normal" code.

In the case I'm concerned with, I used stubs to implement the low-level concurrency primitives that are the basis of the library, so I was wanting to be sure the code with stubs also behaved as expected.

Reply:
Thanks @robdockins , added the scenario you describe as a third reason.